### ISSUE TITLE
Macro expansion can incorrectly bind free variables

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#295]
The following example illustrates that macro expansion does not treat bound variables correctly: `u` is substituted for `x`, turning a free variable into a bound variable. Renaming the bound variable to `y` solves the problem. (Reported by Peter)

```
#!text

field held: Int

function level(this: Ref): Rational

define below(u) (
	(forperm[held] x :: level(x) < level(u))
	// this works: (forperm[held] y :: level(y) < level(u))
)

define acquire(x) {
	exhale below(x)
	inhale acc(x.held)
}

method transfer(this: Ref, to: Ref, amount: Int)
  requires below(this) && level(this) < level(to)
{
  acquire(this)
	acquire(to)
}
```

[Issue created by mschwerhoff: 2016-06-27]
[Last updated on bitbucket: 2017-09-24]

[Comment created by mschwerhoff: 2017-09-24]
No longer reproducible, has probably been solved by Simon Fritsche's master's thesis (AST rewriter framework).

**Comment by fpoli** [comment]
> **@OmerSakar** commented on 2019-10-16 15:02

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2019-10-16 16:21

Thanks for the report and clear repro! I’ve filed an issue for the specific backend [here.](https://bitbucket.org/viperproject/silicon/issues/400/take-drop-with-a-negative-first-index)

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2019-12-23 21:33

Resolved, see [Silicon issue #400](https://bitbucket.org/viperproject/silicon/issues/400).

**Comment by fpoli** [comment]
> **@OmerSakar** commented on 2019-12-24 11:48

Great, thank you for the message.
